orderer/consensus/etcdraft/blockpuller.go

```diff
@@ -107,3 +119,108 @@ func NewBlockPuller(support consensus.ConsenterSupport,
 	}, nil
 }
+
+// NewBlockFetcher creates a new block fetcher
+func NewBlockFetcher(support consensus.ConsenterSupport,
+	baseDialer *cluster.PredicateDialer,
+	clusterConfig localconfig.Cluster,
+	bccsp bccsp.BCCSP,
+) (BlockPuller, error) {
+	verifyBlockSequence := func(blocks []*common.Block, _ string) error {
+		return cluster.VerifyBlocks(blocks, support)
+	}
+
+	stdDialer := &cluster.StandardDialer{
+		Config: baseDialer.Config,
+	}
+	stdDialer.Config.AsyncConnect = false
+	stdDialer.Config.SecOpts.VerifyCertificate = nil
+
+	// Extract the TLS CA certs and endpoints from the configuration,
+	endpoints, err := EndpointconfigFromSupport(support, bccsp)
+	if err != nil {
+		return nil, err
+	}
+
+	der, _ := pem.Decode(stdDialer.Config.SecOpts.Certificate)
+	if der == nil {
+		return nil, errors.Errorf("client certificate isn't in PEM format: %v",
+			string(stdDialer.Config.SecOpts.Certificate))
+	}
+
+	fc := cluster.FetcherConfig{
+		Channel:      support.ChannelID(),
+		Signer:       support,
+		TLSCert:      der.Bytes,
+		Dialer:       stdDialer,
+		Endpoints:    endpoints,
+		FetchTimeout: clusterConfig.ReplicationPullTimeout,
+	}
+
+	// To tolerate byzantine behaviour of `f` faulty nodes, we need a total of `3f + 1` nodes.
+	// f = maxByzantineNodes, total = len(endpoints)
+	maxByzantineNodes := uint64(len(endpoints)-1) / 3
+
+	// timeout for time based shuffling
+	// 10 times the fetchtimeout
+	shuffleTimeout := shuffleTimeoutMultiplier * clusterConfig.ReplicationPullTimeout
+	shuffleTimeoutThrehold := shuffleTimeoutPercentage
+
+	bf := cluster.BlockFetcher{
+		MaxPullBlockRetries: 0,
+		AttestationSourceFactory: func(c cluster.FetcherConfig) cluster.AttestationSource {
+			return &cluster.AttestationPuller{
+				Config: fc,
+				Logger: flogging.MustGetLogger("orderer.common.cluster.attestationpuller").With("channel", support.ChannelID()),
+			}
+		},
+		BlockSourceFactory: func(c cluster.FetcherConfig) cluster.BlockSource {
+			return &cluster.BlockPuller{VerifyBlockSequence: verifyBlockSequence,
+				Logger:              flogging.MustGetLogger("orderer.common.cluster.puller").With("channel", support.ChannelID()),
+				RetryTimeout:        clusterConfig.ReplicationRetryTimeout,
+				MaxTotalBufferBytes: clusterConfig.ReplicationBufferSize,
+				FetchTimeout:        clusterConfig.ReplicationPullTimeout,
+				Endpoints:           c.Endpoints,
+				Signer:              c.Signer,
+				TLSCert:             der.Bytes,
+				Channel:             c.Channel,
+				Dialer:              stdDialer,
+				StopChannel:         make(chan struct{}),
+			}
+		},
+		Config:            fc,
+		Logger:            flogging.MustGetLogger("orderer.common.cluster.puller").With("channel", support.ChannelID()),
+		ShuffleTimeout:    shuffleTimeout,
+		LastShuffledAt:    time.Now(),
+		MaxByzantineNodes: maxByzantineNodes,
+		ConfirmByzantineBehavior: func(attestations []*orderer.BlockAttestation) bool {
+			if attestations == nil {
+				return true
```

Comment:
why true? I thought we return whether we detected byzantine behavior... if we have no attestations then why do we return true?